File: rp2350-reference/split-datasheet.go
package main

import (
	"bytes"
	"errors"
	"fmt"
	"io/fs"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

type pageRange struct {
	name  string
	start int
	end   int
}

type bookmark struct {
	title string
	page  int
}

type alignmentCheck struct {
	rangeName string
	pattern   string
}

var printedRanges = []pageRange{
	{"01-introduction", 13, 23},
	{"02-system-bus", 24, 34},
	{"03a-processor-SIO", 35, 81},
	{"03b-processor-interrupts-debug", 82, 99},
	{"03c-cortex-m33-coprocessors", 100, 122},
	{"03d-cortex-m33-processor", 123, 232},
	{"03e-hazard3-processor", 233, 334},
	{"03f-architecture-switching", 335, 336},
	{"04-memory", 337, 352},
	{"05a-bootrom-concepts", 353, 374},
	{"05b-bootrom-apis", 375, 415},
	{"05c-bootrom-usb-uart", 416, 440},
	{"06-power", 441, 493},
	{"07-resets", 494, 512},
	{"08a-clocks-overview", 513, 553},
	{"08b-clocks-oscillators", 554, 586},
	{"09a-gpio-overview", 587, 603},
	{"09b-gpio-io-user-bank", 604, 759},
	{"09c-gpio-io-qspi-pads", 760, 815},
	{"10-security", 816, 875},
	{"11a-pio-overview-model", 876, 888},
	{"11b-pio-instructions", 889, 901},
	{"11c-pio-details-examples", 902, 960},
	{"12a-uart", 961, 982},
	{"12b-i2c", 983, 1045},
	{"12c-spi", 1046, 1065},
	{"12d-adc-temp", 1066, 1075},
	{"12e-pwm", 1076, 1093},
	{"12f-dma", 1094, 1140},
	{"12g-usb", 1141, 1181},
	{"12h-timers-watchdog", 1182, 1201},
	{"12i-hstx", 1202, 1211},
	{"12j-trng-sha256", 1212, 1225},
	{"12k-qspi-qmi", 1226, 1248},
	{"12l-system-control", 1249, 1267},
	{"13-otp", 1268, 1326},
	{"14-electrical", 1327, 1348},
	{"15-appendices", 1349, 1378},
}

var bookmarkChecks = []alignmentCheck{
	{"01-introduction", `^Chapter 1[.] `},
	{"02-system-bus", `^Chapter 2[.] `},
	{"03a-processor-SIO", `^Chapter 3[.] `},
	{"04-memory", `^Chapter 4[.] `},
	{"05a-bootrom-concepts", `^Chapter 5[.] `},
	{"06-power", `^Chapter 6[.] `},
	{"07-resets", `^Chapter 7[.] `},
	{"08a-clocks-overview", `^Chapter 8[.] `},
	{"09a-gpio-overview", `^Chapter 9[.] `},
	{"10-security", `^Chapter 10[.] `},
	{"11a-pio-overview-model", `^Chapter 11[.] `},
	{"12a-uart", `^Chapter 12[.] `},
	{"13-otp", `^Chapter 13[.] `},
	{"14-electrical", `^Chapter 14[.] `},
	{"15-appendices", `^Appendix A:`},
}

func die(format string, args ...any) {
	fmt.Fprintf(os.Stderr, "error: "+format+"\n", args...)
	os.Exit(1)
}

func warn(format string, args ...any) {
	fmt.Fprintf(os.Stderr, "warning: "+format+"\n", args...)
}

func requireCommands(names ...string) {
	for _, name := range names {
		if _, err := exec.LookPath(name); err != nil {
			die("%s is required but not installed", name)
		}
	}
}

func runCapture(command string, args ...string) string {
	cmd := exec.Command(command, args...)
	var stderr bytes.Buffer
	cmd.Stderr = &stderr
	output, err := cmd.Output()
	if err != nil {
		details := strings.TrimSpace(stderr.String())
		if details == "" {
			details = err.Error()
		}
		die("command failed: %s %s (%s)", command, strings.Join(args, " "), details)
	}
	return string(output)
}

func run(command string, args ...string) {
	cmd := exec.Command(command, args...)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	if err := cmd.Run(); err != nil {
		die("command failed: %s %s (%v)", command, strings.Join(args, " "), err)
	}
}

func parseBookmarks(dump string) []bookmark {
	var bookmarks []bookmark
	currentTitle := ""

	for _, line := range strings.Split(dump, "\n") {
		trimmed := strings.TrimSpace(line)
		if strings.HasPrefix(trimmed, "BookmarkTitle: ") {
			currentTitle = strings.TrimPrefix(trimmed, "BookmarkTitle: ")
			continue
		}

		if strings.HasPrefix(trimmed, "BookmarkPageNumber: ") && currentTitle != "" {
			pageText := strings.TrimPrefix(trimmed, "BookmarkPageNumber: ")
			pageNumber, err := strconv.Atoi(pageText)
			if err != nil {
				die("invalid bookmark page number: %s", pageText)
			}
			bookmarks = append(bookmarks, bookmark{title: currentTitle, page: pageNumber})
			currentTitle = ""
		}
	}

	if len(bookmarks) == 0 {
		die("no bookmarks found in PDF metadata")
	}

	return bookmarks
}

func bookmarkPage(bookmarks []bookmark, pattern string) int {
	compiled := regexp.MustCompile(pattern)
	for _, item := range bookmarks {
		if compiled.MatchString(item.title) {
			return item.page
		}
	}
	die("could not find bookmark matching regex: %s", pattern)
	return 0
}

func rangeStart(ranges []pageRange, name string) (int, bool) {
	for _, item := range ranges {
		if item.name == name {
			return item.start, true
		}
	}
	return 0, false
}

func buildPhysicalRanges(bookmarks []bookmark) ([]pageRange, int) {
	chapter1Start := bookmarkPage(bookmarks, `^Chapter 1[.] `)
	appendixStart := bookmarkPage(bookmarks, `^Appendix A: `)

	pageOffset := chapter1Start - 13
	if pageOffset < 0 {
		die("calculated negative page offset: %d", pageOffset)
	}

	frontMatterEnd := chapter1Start - 1
	if frontMatterEnd < 1 {
		die("invalid front matter end page: %d", frontMatterEnd)
	}

	physicalRanges := []pageRange{{name: "00-front-matter", start: 1, end: frontMatterEnd}}
	for _, item := range printedRanges {
		physicalRanges = append(physicalRanges, pageRange{
			name:  item.name,
			start: item.start + pageOffset,
			end:   item.end + pageOffset,
		})
	}

	appendicesRangeStart, found := rangeStart(physicalRanges, "15-appendices")
	if !found {
		die("missing generated range for 15-appendices")
	}
	if appendicesRangeStart != appendixStart {
		die(
			"appendix start mismatch: range starts at %d, bookmark starts at %d",
			appendicesRangeStart,
			appendixStart,
		)
	}

	return physicalRanges, pageOffset
}

func validateRanges(ranges []pageRange, totalPages int) {
	previousEnd := 0
	for lineNumber, item := range ranges {
		displayLine := lineNumber + 1
		if item.start > item.end {
			die("line %d has start > end: %s %d-%d", displayLine, item.name, item.start, item.end)
		}
		if item.start <= previousEnd {
			die(
				"line %d overlaps prior range: %s starts at %d, previous ended at %d",
				displayLine, item.name, item.start, previousEnd,
			)
		}
		if item.start != previousEnd+1 {
			die(
				"line %d has gap before %s: expected %d, got %d",
				displayLine, item.name, previousEnd+1, item.start,
			)
		}
		if item.end > totalPages {
			die(
				"line %d exceeds source page count (%d): %s ends at %d",
				displayLine, totalPages, item.name, item.end,
			)
		}
		previousEnd = item.end
	}

	if previousEnd < totalPages {
		warn(
			"ranges end at %d but source has %d pages (trailing pages excluded)",
			previousEnd, totalPages,
		)
	}
}

func validateBookmarkAlignment(ranges []pageRange, bookmarks []bookmark) {
	for _, check := range bookmarkChecks {
		expected := bookmarkPage(bookmarks, check.pattern)
		actual, found := rangeStart(ranges, check.rangeName)
		if !found {
			die("could not find generated range: %s", check.rangeName)
		}
		if actual != expected {
			die(
				"bookmark mismatch for %s: range starts at %d, bookmark starts at %d",
				check.rangeName, actual, expected,
			)
		}
	}
}

func splitRange(srcPath, outDir string, item pageRange) {
	pages := item.end - item.start + 1
	fmt.Printf("  %s (%dpp: %d-%d)\n", item.name, pages, item.start, item.end)
	outputPath := filepath.Join(outDir, item.name+".pdf")
	run(
		"qpdf",
		srcPath,
		"--pages",
		".",
		fmt.Sprintf("%d-%d", item.start, item.end),
		"--",
		outputPath,
	)
}

func humanSize(sizeBytes int64) string {
	units := []string{"B", "K", "M", "G", "T"}
	size := float64(sizeBytes)
	unitIndex := 0
	for size >= 1024 && unitIndex < len(units)-1 {
		size /= 1024
		unitIndex++
	}
	if unitIndex == 0 {
		return fmt.Sprintf("%d%s", int(size), units[unitIndex])
	}
	return fmt.Sprintf("%.1f%s", size, units[unitIndex])
}

func totalPages(srcPath string) int {
	text := strings.TrimSpace(runCapture("qpdf", "--show-npages", srcPath))
	pages, err := strconv.Atoi(text)
	if err != nil {
		die("invalid total page count from qpdf: %s", text)
	}
	return pages
}

func defaultBaseDir() string {
	candidates := []string{}

	executablePath, execErr := os.Executable()
	if execErr == nil {
		candidates = append(candidates, filepath.Dir(executablePath))
	}
	currentDir, cwdErr := os.Getwd()
	if cwdErr == nil {
		candidates = append(candidates, currentDir, filepath.Join(currentDir, "rp2350-reference"))
	}

	for _, candidate := range candidates {
		pdfPath := filepath.Join(candidate, "rp2350-datasheet.pdf")
		if _, err := os.Stat(pdfPath); err == nil {
			return candidate
		}
	}

	if len(candidates) > 0 {
		return candidates[0]
	}
	return "."
}

func listAndPrintOutputPDFs(outDir string) {
	entries := []struct {
		name string
		size int64
	}{}

	filePattern := filepath.Join(outDir, "*.pdf")
	matches, globErr := filepath.Glob(filePattern)
	if globErr != nil {
		die("failed to list output PDFs: %v", globErr)
	}
	for _, match := range matches {
		info, statErr := os.Stat(match)
		if statErr != nil {
			if errors.Is(statErr, fs.ErrNotExist) {
				continue
			}
			die("failed to stat output PDF %s: %v", match, statErr)
		}
		entries = append(entries, struct {
			name string
			size int64
		}{
			name: filepath.Base(match),
			size: info.Size(),
		})
	}

	sort.Slice(entries, func(first, second int) bool {
		return entries[first].size > entries[second].size
	})

	for _, entry := range entries {
		fmt.Printf("%5s %s\n", humanSize(entry.size), entry.name)
	}
}

func main() {
	if len(os.Args) > 3 {
		fmt.Fprintf(os.Stderr, "usage: %s [source.pdf] [output_dir]\n", filepath.Base(os.Args[0]))
		os.Exit(2)
	}

	baseDir := defaultBaseDir()
	srcPath := filepath.Join(baseDir, "rp2350-datasheet.pdf")
	outDir := filepath.Join(baseDir, "datasheet")

	if len(os.Args) >= 2 {
		srcPath = os.Args[1]
	}
	if len(os.Args) >= 3 {
		outDir = os.Args[2]
	}

	requireCommands("qpdf", "pdftk")
	if _, err := os.Stat(srcPath); err != nil {
		die("source PDF not found: %s", srcPath)
	}

	bookmarkDump := runCapture("pdftk", srcPath, "dump_data")
	bookmarks := parseBookmarks(bookmarkDump)
	ranges, pageOffset := buildPhysicalRanges(bookmarks)
	validateRanges(ranges, totalPages(srcPath))
	validateBookmarkAlignment(ranges, bookmarks)

	if err := os.MkdirAll(outDir, 0o755); err != nil {
		die("failed to create output directory %s: %v", outDir, err)
	}

	fmt.Printf("Splitting RP2350 datasheet from: %s (page offset: +%d)\n\n", srcPath, pageOffset)
	for _, item := range ranges {
		splitRange(srcPath, outDir, item)
	}

	fmt.Printf("\nDone! Files in: %s\n", outDir)
	listAndPrintOutputPDFs(outDir)
}

File: rp2350-reference/split-datasheet.rs
use std::cmp::Reverse;
use std::env;
use std::ffi::OsStr;
use std::fs;
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};

#[derive(Clone, Debug)]
struct PageRange {
    name: &'static str,
    start: i32,
    end: i32,
}

#[derive(Clone, Debug)]
struct Bookmark {
    title: String,
    page: i32,
}

#[derive(Clone, Debug)]
struct BookmarkCheck {
    range_name: &'static str,
    title_prefix: &'static str,
}

fn printed_ranges() -> Vec<PageRange> {
    vec![
        PageRange {
            name: "01-introduction",
            start: 13,
            end: 23,
        },
        PageRange {
            name: "02-system-bus",
            start: 24,
            end: 34,
        },
        PageRange {
            name: "03a-processor-SIO",
            start: 35,
            end: 81,
        },
        PageRange {
            name: "03b-processor-interrupts-debug",
            start: 82,
            end: 99,
        },
        PageRange {
            name: "03c-cortex-m33-coprocessors",
            start: 100,
            end: 122,
        },
        PageRange {
            name: "03d-cortex-m33-processor",
            start: 123,
            end: 232,
        },
        PageRange {
            name: "03e-hazard3-processor",
            start: 233,
            end: 334,
        },
        PageRange {
            name: "03f-architecture-switching",
            start: 335,
            end: 336,
        },
        PageRange {
            name: "04-memory",
            start: 337,
            end: 352,
        },
        PageRange {
            name: "05a-bootrom-concepts",
            start: 353,
            end: 374,
        },
        PageRange {
            name: "05b-bootrom-apis",
            start: 375,
            end: 415,
        },
        PageRange {
            name: "05c-bootrom-usb-uart",
            start: 416,
            end: 440,
        },
        PageRange {
            name: "06-power",
            start: 441,
            end: 493,
        },
        PageRange {
            name: "07-resets",
            start: 494,
            end: 512,
        },
        PageRange {
            name: "08a-clocks-overview",
            start: 513,
            end: 553,
        },
        PageRange {
            name: "08b-clocks-oscillators",
            start: 554,
            end: 586,
        },
        PageRange {
            name: "09a-gpio-overview",
            start: 587,
            end: 603,
        },
        PageRange {
            name: "09b-gpio-io-user-bank",
            start: 604,
            end: 759,
        },
        PageRange {
            name: "09c-gpio-io-qspi-pads",
            start: 760,
            end: 815,
        },
        PageRange {
            name: "10-security",
            start: 816,
            end: 875,
        },
        PageRange {
            name: "11a-pio-overview-model",
            start: 876,
            end: 888,
        },
        PageRange {
            name: "11b-pio-instructions",
            start: 889,
            end: 901,
        },
        PageRange {
            name: "11c-pio-details-examples",
            start: 902,
            end: 960,
        },
        PageRange {
            name: "12a-uart",
            start: 961,
            end: 982,
        },
        PageRange {
            name: "12b-i2c",
            start: 983,
            end: 1045,
        },
        PageRange {
            name: "12c-spi",
            start: 1046,
            end: 1065,
        },
        PageRange {
            name: "12d-adc-temp",
            start: 1066,
            end: 1075,
        },
        PageRange {
            name: "12e-pwm",
            start: 1076,
            end: 1093,
        },
        PageRange {
            name: "12f-dma",
            start: 1094,
            end: 1140,
        },
        PageRange {
            name: "12g-usb",
            start: 1141,
            end: 1181,
        },
        PageRange {
            name: "12h-timers-watchdog",
            start: 1182,
            end: 1201,
        },
        PageRange {
            name: "12i-hstx",
            start: 1202,
            end: 1211,
        },
        PageRange {
            name: "12j-trng-sha256",
            start: 1212,
            end: 1225,
        },
        PageRange {
            name: "12k-qspi-qmi",
            start: 1226,
            end: 1248,
        },
        PageRange {
            name: "12l-system-control",
            start: 1249,
            end: 1267,
        },
        PageRange {
            name: "13-otp",
            start: 1268,
            end: 1326,
        },
        PageRange {
            name: "14-electrical",
            start: 1327,
            end: 1348,
        },
        PageRange {
            name: "15-appendices",
            start: 1349,
            end: 1378,
        },
    ]
}

fn bookmark_checks() -> Vec<BookmarkCheck> {
    vec![
        BookmarkCheck {
            range_name: "01-introduction",
            title_prefix: "Chapter 1. ",
        },
        BookmarkCheck {
            range_name: "02-system-bus",
            title_prefix: "Chapter 2. ",
        },
        BookmarkCheck {
            range_name: "03a-processor-SIO",
            title_prefix: "Chapter 3. ",
        },
        BookmarkCheck {
            range_name: "04-memory",
            title_prefix: "Chapter 4. ",
        },
        BookmarkCheck {
            range_name: "05a-bootrom-concepts",
            title_prefix: "Chapter 5. ",
        },
        BookmarkCheck {
            range_name: "06-power",
            title_prefix: "Chapter 6. ",
        },
        BookmarkCheck {
            range_name: "07-resets",
            title_prefix: "Chapter 7. ",
        },
        BookmarkCheck {
            range_name: "08a-clocks-overview",
            title_prefix: "Chapter 8. ",
        },
        BookmarkCheck {
            range_name: "09a-gpio-overview",
            title_prefix: "Chapter 9. ",
        },
        BookmarkCheck {
            range_name: "10-security",
            title_prefix: "Chapter 10. ",
        },
        BookmarkCheck {
            range_name: "11a-pio-overview-model",
            title_prefix: "Chapter 11. ",
        },
        BookmarkCheck {
            range_name: "12a-uart",
            title_prefix: "Chapter 12. ",
        },
        BookmarkCheck {
            range_name: "13-otp",
            title_prefix: "Chapter 13. ",
        },
        BookmarkCheck {
            range_name: "14-electrical",
            title_prefix: "Chapter 14. ",
        },
        BookmarkCheck {
            range_name: "15-appendices",
            title_prefix: "Appendix A:",
        },
    ]
}

fn die(message: &str) -> ! {
    eprintln!("error: {message}");
    std::process::exit(1);
}

fn warn(message: &str) {
    eprintln!("warning: {message}");
}

fn require_commands(commands: &[&str]) {
    for command in commands {
        if Command::new(command)
            .arg("--version")
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status()
            .is_err()
        {
            die(&format!("{command} is required but not installed"));
        }
    }
}

fn run_capture(command: &str, args: &[&str]) -> String {
    let output = Command::new(command).args(args).output().unwrap_or_else(|error| {
        die(&format!(
            "failed to execute command: {command} {} ({error})",
            args.join(" ")
        ));
    });
    if !output.status.success() {
        let stderr_text = String::from_utf8_lossy(&output.stderr).trim().to_string();
        let details = if stderr_text.is_empty() {
            format!("exit {}", output.status)
        } else {
            stderr_text
        };
        die(&format!(
            "command failed: {command} {} ({details})",
            args.join(" ")
        ));
    }
    String::from_utf8_lossy(&output.stdout).to_string()
}

fn run(command: &str, args: &[&str]) {
    let status = Command::new(command)
        .args(args)
        .status()
        .unwrap_or_else(|error| {
            die(&format!(
                "failed to execute command: {command} {} ({error})",
                args.join(" ")
            ));
        });
    if !status.success() {
        die(&format!(
            "command failed: {command} {} (exit {status})",
            args.join(" ")
        ));
    }
}

fn parse_bookmarks(bookmark_dump: &str) -> Vec<Bookmark> {
    let mut current_title: Option<String> = None;
    let mut bookmarks: Vec<Bookmark> = Vec::new();

    for line in bookmark_dump.lines() {
        let trimmed = line.trim();
        if let Some(title) = trimmed.strip_prefix("BookmarkTitle: ") {
            current_title = Some(title.to_string());
            continue;
        }

        if let Some(page_text) = trimmed.strip_prefix("BookmarkPageNumber: ") {
            if let Some(title) = current_title.take() {
                let page = page_text.parse::<i32>().unwrap_or_else(|_| {
                    die(&format!("invalid bookmark page number: {page_text}"));
                });
                bookmarks.push(Bookmark { title, page });
            }
        }
    }

    if bookmarks.is_empty() {
        die("no bookmarks found in PDF metadata");
    }

    bookmarks
}

fn bookmark_page(bookmarks: &[Bookmark], prefix: &str) -> i32 {
    for bookmark in bookmarks {
        if bookmark.title.starts_with(prefix) {
            return bookmark.page;
        }
    }
    die(&format!(
        "could not find bookmark starting with: {prefix}"
    ));
}

fn range_start(ranges: &[PageRange], range_name: &str) -> Option<i32> {
    ranges
        .iter()
        .find(|item| item.name == range_name)
        .map(|item| item.start)
}

fn build_physical_ranges(bookmarks: &[Bookmark]) -> (Vec<PageRange>, i32) {
    let chapter1_start = bookmark_page(bookmarks, "Chapter 1. ");
    let appendix_start = bookmark_page(bookmarks, "Appendix A:");

    let page_offset = chapter1_start - 13;
    if page_offset < 0 {
        die(&format!("calculated negative page offset: {page_offset}"));
    }

    let front_matter_end = chapter1_start - 1;
    if front_matter_end < 1 {
        die(&format!("invalid front matter end page: {front_matter_end}"));
    }

    let mut physical_ranges = vec![PageRange {
        name: "00-front-matter",
        start: 1,
        end: front_matter_end,
    }];

    for range in printed_ranges() {
        physical_ranges.push(PageRange {
            name: range.name,
            start: range.start + page_offset,
            end: range.end + page_offset,
        });
    }

    let appendices_start = range_start(&physical_ranges, "15-appendices")
        .unwrap_or_else(|| die("missing generated range: 15-appendices"));
    if appendices_start != appendix_start {
        die(&format!(
            "appendix start mismatch: range starts at {appendices_start}, bookmark starts at {appendix_start}"
        ));
    }

    (physical_ranges, page_offset)
}

fn validate_ranges(ranges: &[PageRange], total_pages: i32) {
    let mut previous_end = 0;
    for (index, range) in ranges.iter().enumerate() {
        let line_number = index + 1;
        if range.start > range.end {
            die(&format!(
                "line {line_number} has start > end: {} {}-{}",
                range.name, range.start, range.end
            ));
        }
        if range.start <= previous_end {
            die(&format!(
                "line {line_number} overlaps prior range: {} starts at {}, previous ended at {}",
                range.name, range.start, previous_end
            ));
        }
        if range.start != previous_end + 1 {
            die(&format!(
                "line {line_number} has gap before {}: expected {}, got {}",
                range.name,
                previous_end + 1,
                range.start
            ));
        }
        if range.end > total_pages {
            die(&format!(
                "line {line_number} exceeds source page count ({total_pages}): {} ends at {}",
                range.name, range.end
            ));
        }
        previous_end = range.end;
    }

    if previous_end < total_pages {
        warn(&format!(
            "ranges end at {previous_end} but source has {total_pages} pages (trailing pages excluded)"
        ));
    }
}

fn validate_bookmark_alignment(ranges: &[PageRange], bookmarks: &[Bookmark]) {
    for check in bookmark_checks() {
        let expected = bookmark_page(bookmarks, check.title_prefix);
        let actual = range_start(ranges, check.range_name)
            .unwrap_or_else(|| die(&format!("could not find generated range: {}", check.range_name)));
        if actual != expected {
            die(&format!(
                "bookmark mismatch for {}: range starts at {actual}, bookmark starts at {expected}",
                check.range_name
            ));
        }
    }
}

fn split_range(src_path: &Path, out_dir: &Path, range: &PageRange) {
    let pages = range.end - range.start + 1;
    println!("  {} ({}pp: {}-{})", range.name, pages, range.start, range.end);
    let output_path = out_dir.join(format!("{}.pdf", range.name));
    let page_spec = format!("{}-{}", range.start, range.end);
    run(
        "qpdf",
        &[
            src_path.to_string_lossy().as_ref(),
            "--pages",
            ".",
            page_spec.as_str(),
            "--",
            output_path.to_string_lossy().as_ref(),
        ],
    );
}

fn human_size(size_bytes: u64) -> String {
    let units = ["B", "K", "M", "G", "T"];
    let mut size = size_bytes as f64;
    let mut unit_index = 0usize;
    while size >= 1024.0 && unit_index < units.len() - 1 {
        size /= 1024.0;
        unit_index += 1;
    }
    if unit_index == 0 {
        format!("{}{}", size as u64, units[unit_index])
    } else {
        format!("{size:.1}{}", units[unit_index])
    }
}

fn default_base_dir() -> PathBuf {
    let mut candidates: Vec<PathBuf> = Vec::new();

    if let Ok(executable_path) = env::current_exe() {
        if let Some(parent) = executable_path.parent() {
            candidates.push(parent.to_path_buf());
        }
    }

    if let Ok(current_dir) = env::current_dir() {
        candidates.push(current_dir.clone());
        candidates.push(current_dir.join("rp2350-reference"));
    }

    for candidate in &candidates {
        if candidate.join("rp2350-datasheet.pdf").is_file() {
            return candidate.clone();
        }
    }

    candidates.into_iter().next().unwrap_or_else(|| PathBuf::from("."))
}

fn total_pages(src_path: &Path) -> i32 {
    let output = run_capture("qpdf", &["--show-npages", src_path.to_string_lossy().as_ref()]);
    output.trim().parse::<i32>().unwrap_or_else(|_| {
        die(&format!("invalid total page count from qpdf: {}", output.trim()))
    })
}

fn list_output_pdfs(out_dir: &Path) {
    let mut entries: Vec<(String, u64)> = Vec::new();
    let read_dir = fs::read_dir(out_dir).unwrap_or_else(|error| {
        die(&format!(
            "failed to list output directory {}: {error}",
            out_dir.to_string_lossy()
        ))
    });

    for result in read_dir {
        let entry = result.unwrap_or_else(|error| {
            die(&format!(
                "failed to read output directory entry in {}: {error}",
                out_dir.to_string_lossy()
            ))
        });
        let path = entry.path();
        if path.extension() != Some(OsStr::new("pdf")) {
            continue;
        }
        let metadata = entry.metadata().unwrap_or_else(|error| {
            die(&format!(
                "failed to stat output file {}: {error}",
                path.to_string_lossy()
            ))
        });
        entries.push((entry.file_name().to_string_lossy().to_string(), metadata.len()));
    }

    entries.sort_by_key(|item| Reverse(item.1));
    for (name, size) in entries {
        println!("{:>5} {}", human_size(size), name);
    }
}

fn main() {
    let arguments: Vec<String> = env::args().collect();
    if arguments.len() > 3 {
        let binary_name = Path::new(&arguments[0])
            .file_name()
            .and_then(|name| name.to_str())
            .unwrap_or("split-datasheet.rs");
        eprintln!("usage: {binary_name} [source.pdf] [output_dir]");
        std::process::exit(2);
    }

    let base_dir = default_base_dir();
    let mut src_path = base_dir.join("rp2350-datasheet.pdf");
    let mut out_dir = base_dir.join("datasheet");

    if arguments.len() >= 2 {
        src_path = PathBuf::from(&arguments[1]);
    }
    if arguments.len() >= 3 {
        out_dir = PathBuf::from(&arguments[2]);
    }

    require_commands(&["qpdf", "pdftk"]);
    if !src_path.is_file() {
        die(&format!(
            "source PDF not found: {}",
            src_path.to_string_lossy()
        ));
    }

    let bookmark_dump = run_capture("pdftk", &[src_path.to_string_lossy().as_ref(), "dump_data"]);
    let bookmarks = parse_bookmarks(&bookmark_dump);
    let (ranges, page_offset) = build_physical_ranges(&bookmarks);
    validate_ranges(&ranges, total_pages(&src_path));
    validate_bookmark_alignment(&ranges, &bookmarks);

    fs::create_dir_all(&out_dir).unwrap_or_else(|error| {
        die(&format!(
            "failed to create output directory {}: {error}",
            out_dir.to_string_lossy()
        ))
    });

    println!(
        "Splitting RP2350 datasheet from: {} (page offset: +{})",
        src_path.to_string_lossy(),
        page_offset
    );
    println!();

    for range in &ranges {
        split_range(&src_path, &out_dir, range);
    }

    println!();
    println!("Done! Files in: {}", out_dir.to_string_lossy());
    list_output_pdfs(&out_dir);
}
